### ISSUE TITLE
lxc-execute failed to exec a command in unprivileged mode on Ubuntu 14.04

### ISSUE THREAD

**Issue by sartrevan** [lxc/lxc#417]
I'm trying to use lxc to start a java process in unprivileged mode following the guide:https://www.stgraber.org/2014/01/17/lxc-1-0-unprivileged-containers/

I've managed to start an lxc ubuntu guest os but I'd rather use lxc to start a java process of my own as an application container. Is it possible? I've googled for 2 days, still no luck.

![image](https://cloud.githubusercontent.com/assets/4744114/5898826/6251befa-a58e-11e4-8aa6-b7730db888bf.png)

**Comment by stgraber** [comment]
What distro, kernel and version of LXC is that?

**Comment by sartrevan** [comment]
**Distro**
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.1 LTS"
NAME="Ubuntu"
VERSION="14.04.1 LTS, Trusty Tahr"
**Kernel**
3.13.0-32-generic
**LXC Version**
1.0.7

**Comment by ptoomey3** [comment]
I ran into what seems like a directly related issue when playing with `lxc-execute` and unprivileged containers. It seems that `lxc-execute` assumes that you are either running a privileged container or are starting an unprivileged container where root is mapped to some subordinate IDs. It looks like `lxc-execute` always does a setuid(0) in https://github.com/lxc/lxc/blob/3b6aebceaf331dd4967f86af0632a3b1cde0120c/src/lxc/start.c#L672 when you invoke a new namespace. It isn't clear to me that calling `lxc-execute`, as the original submitter did  above (without a uid/gid mapping) automatically creates a new user namespace or if that wouldn't happen unless you passed an explicit uid/gid map config.

Regardless of the above, it seems like `lxc-execute` doesn't work as one would think it should when you are not running as root. Two quick examples:

```
lxc-execute -n foobar -- whoami
lxc-execute: start.c: must_drop_cap_sys_boot: 554 failed to clone (0x20000011): Operation not permitted
lxc-execute: namespace.c: lxc_clone: 67 failed to clone (0x6c020000): Operation not permitted
lxc-execute: start.c: lxc_spawn: 897 Operation not permitted - failed to fork into a new namespace
lxc-execute: start.c: __lxc_start: 1087 failed to spawn 'foobar'
```

In that case the command fails to run at all. If I explicitly pass a uid/gid map like below (it maps my host uid to the container uid and maps the host root to a subordinate ID)  I at least get the command to execute:

```
cat .config/lxc/default.conf
lxc.id_map = u 1000 100001 1
lxc.id_map = u 0 100000 1
lxc.id_map = g 0 100000 1
lxc.id_map = g 1000 100001 1
```

```
lxc-execute -n foobar -f ~/.config/lxc/default.conf -- whoami
lxc-execute: conf.c: lxc_execute_bind_init: 3925 Permission denied - Failed to create mount target '/usr/lib/x86_64-linux-gnu/lxc/init.lxc.static'
lxc: cgmanager.c: lxc_cgmanager_escape: 314 call to cgmanager_move_pid_abs_sync(name=systemd) failed: Escape request from different namespace requires a proxy
init.lxc: log.c: log_open: 184 failed to open log file "/home/ptoomey3/.local/share/lxc/foobar/foobar.log" : Permission denied
init.lxc: log.c: log_open: 184 failed to open log file "/var/log/lxc/foobar.log" : Permission denied
root
```

In this case I get a few errors but I at least get the process to execute. But, it is run as container uid 0. This seems pretty unexpected. If I mapped my host uid 1000 to the container uid 1000 I would expect `whoami` to return the user that executed `lxc-execute`. It seems like this may be an artifact of the fact that `init.lxc` is used to launch the passed in executable. It looks like `init.lxc` is set to uid 0 in https://github.com/lxc/lxc/blob/3b6aebceaf331dd4967f86af0632a3b1cde0120c/src/lxc/start.c#L672. This makes sense from the "full container" perspective, since init is always uid 0 and is responsible for staring up the system. But, for the case of using `lxc-execute` to invoke a host command using "default isolation", it seems unexpected. So, I wonder if it should detect the context it is being invoked from and do the right thing. I made a quick change to https://github.com/lxc/lxc/blob/3b6aebceaf331dd4967f86af0632a3b1cde0120c/src/lxc/start.c#L672 and just commented out the line. That ended up doing what I expected. It successfully launched `whoami` with my original uid being returned from within the container. But, given that the `setuid(0)` [has a comment](https://github.com/lxc/lxc/blob/3b6aebceaf331dd4967f86af0632a3b1cde0120c/src/lxc/start.c#L663-L664) that implies this call is required for the parent to have "privilege over our namespace", I don't want to assume that the change is that simple. In other words, I don't have a full appreciation for what code after that invocation assumes/relies on the parent being uid 0. 

So, what do you all think would be necessary for `lxc-execute` to properly support unprivileged containers? My expectation would be that a non root user should be able to execute `lxc-execute -n foobar -- some_utility` and have it execute as their UID with some default isolation applied. If the changes necessary to support unprivileged `lxc-execute` are fairly targeted I could take a pass at making the updates. I was hesitant to start poking around without someone more familiar with the codebase providing some guidance on what would need to be updated.

**Comment by ptoomey3** [comment]
@hallyn @stgraber - Any thoughts on the above? If not, do you know of anyone else to CC on this issue to narrow down the bits of code that would need updating to make this work?

**Comment by hallyn** [comment]
The lxc-devel list would be the place to discuss this.  If anyone is interested, they would pipe in there.

As for the comment about needing privilege over our namespace - well, yes, you need to become root before any exec if you want privilege.  If you don't want privilege - which you do not - then it doesn't matter.

If you make the change to start.c:672 so that it only tries to setgid(0) and setuid(0) if gid/uid 0, respectively, are mapped in the container's user namespace, then if that works for you please sent the patch and I'll ack it.

**Comment by ptoomey3** [comment]
Thanks @hallyn - I copied my comment above (with small modifications) and submitted it to the lxc-devel mailing list.

**Comment by ptoomey3** [comment]
@stgraber - I'm not 100% sure whether https://github.com/lxc/lxc/pull/585 fixes the issue the original submitter was seeing. But, in theory, that PR should allow him to execute a process in an unprivileged container under his original UID.

**Comment by brauner** [comment]
This should be fixed via d08f8d2f74c7407c40fc80e3878d631bd8f50214 .
